copy/copy.go

```diff
@@ -655,17 +696,108 @@ func (ic *imageCopier) copyLayer(ctx context.Context, srcInfo types.BlobInfo, po
 	cachedDiffID := ic.c.blobInfoCache.UncompressedDigest(srcInfo.Digest) // May be ""
 	diffIDIsNeeded := ic.diffIDsAreNeeded && cachedDiffID == ""
 
+	// Create the initial placeholder bar which is later being replaced by
+	// another one.
+	bar := ic.c.createProgressBar(pool, srcInfo, "blob", " ", mpb.BarRemoveOnComplete())
+
 	// If we already have the blob, and we don't need to compute the diffID, then we don't need to read it from the source.
 	if !diffIDIsNeeded {
-		reused, blobInfo, err := ic.c.dest.TryReusingBlob(ctx, srcInfo, ic.c.blobInfoCache, ic.canSubstituteBlobs)
+		// tryReusingBlob is an anonymous function to avoid adding
+		// another method with a complex parameter list and semantics.
+		// It's sole purpose is to simplify the control with an early
+		// return if the image destination does not blob locking.
+		tryReusingBlob := func() (bool, types.BlobInfo, digest.Digest, error) {
+			reused, blobInfo, err := ic.c.dest.TryReusingBlob(ctx, srcInfo, ic.c.blobInfoCache, ic.canSubstituteBlobs)
+			if err != nil {
+				return false, types.BlobInfo{}, "", errors.Wrapf(err, "Error trying to reuse blob %s at destination", srcInfo.Digest)
+			}
+			if reused {
+				logrus.Debugf("Skipping blob %s (already present):", srcInfo.Digest)
+				bar.SetTotal(0, true)
+				ic.c.createProgressBar(pool, srcInfo, "blob", "skipped: already exists", mpb.BarReplaceOnComplete(bar))
+				return true, blobInfo, cachedDiffID, nil
+			}
+
+			if !ic.c.dest.SupportsBlobLocks() {
+				return false, types.BlobInfo{}, "", nil
+			}
+
+			// Check if another process might be copying the layer at the
+			// moment.  If so, wait for it's completion and try to reuse
+			// the blob again.
+			//
+			logrus.Debugf("Checking if blob %s is already being copied by another process", srcInfo.Digest)
+
+			// LockBlob() does not support try-lock semantics which would
+			// be a straight-forward mechanism to detect if another process
+			// is currently copying a specific layer.  However, with
+			// LockBlob() we are constrained by the file locks in
+			// containers/storage which cannot support try-lock semantics
+			// as fcntl(2) does not support that.
```

Comment:
`F_SETLK` vs. `F_SETLKW` does support that AFAICS. (The `sync.Mutex` objects used in `c/storage.lockfile` seemingly don’t, OTOH.)

Reply:
The manpage states: _"If a conflicting lock is held by another process, this call returns -1 and sets errno to EACCES or EAGAIN."_

This means that we only get a _EGAIN_ if we change the type of the lock (reader <-> writer). Here we need a write-lock and hence never change the type. In this case, it would block every time.

Reply:
> This means that we only get a EGAIN if we change the type of the lock (reader <-> writer).

FWIW, I don’t think that’s what the man page means; see elsewhere in there
> Conflicting lock combinations (i.e., a read lock and a write lock or two write locks) 

(admittedly in a quite different section).

Compare also fcntl(3p)
> If a shared or exclusive lock cannot be set, fcntl() shall return immediately with a return value of −1.

or (info libc)
> [F_SETLKW] is just like the ‘F_SETLK’ command, but causes the process to block (or wait) until the request can be specified.

---

Anyway, as long as the in-process `rwMutex` does not support the same operation, this does not change much — only maybe the comment could be corrected.